### ISSUE TITLE
[Feature Request] Make MSAL.NET compatible with MAUI 

### ISSUE THREAD

**Issue by SameerK-MSFT** [AzureAD/microsoft-authentication-library-for-dotnet#3127]
Initial PR to add support: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3262

- [x] Add Android and iOS support first
- [x] Fix the CI build as well
- [x] Manual testing Android and iOS scenarios (embedded, system and broker)
- [x] Update xamarin sample

At this point it we should be able to do a partial release with MAUI support for mobile.

- ~[ ] MacCatalyst support (embedded, system and broker (?) scenario)~ CUT. 

MSAL ObjC does not have MacCatalyst support, we will need to coordinate with them on this. Should not block GA of MAUI mobile targets.

- [x] Net6 (WinUI) support (embedded, system, broker)
- [x] Update sample ~with MacCatalyst and~ WinUI
- [ ] Automation tests for ios and android

Maui Status can be found here:
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/MauiStatus.md

**Comment by bgavrilMS** [comment]
Blocked by MAUI previews

**Comment by xperiandri** [comment]
Could you release a preview package targeting `net6.0-android`/`net6.0-ios`/`net6.0-macos`/`net6.0-maccatalyst`?
Just add that targets and create an appropriate package

**Comment by xperiandri** [comment]
Build with latest preview

**Comment by bgavrilMS** [comment]
.net6.0-android should be compatible with our current target.
.net6.0-ios is undergoing breaking changes and we will need to fix later on. Currently broken.
We have not yet tested with mac, but should work.

**Comment by ViktorArbuzov** [comment]
using this package into MAUI project, can't deploy to remote mac because of this

**Comment by bgavrilMS** [comment]
Yes, there is PR out which provides some support. We've got some trouble on MacCatalyst and also we need to do more testing.

**Comment by igalfsg** [comment]
@SameerK-MSFT @bgavrilMS is there an update on this feature?

**Comment by bgavrilMS** [comment]
We're hoping to release mobile support soon. Problem we're facing is that ecosystem (VS, AzureDevOps agents etc.) do not support it natively, need to do some build work.

**Comment by lnaie** [comment]
Exciting. When is it gonna be dished out?

**Comment by Jeremywhiteley** [comment]
Any idea on when this will be fixed?

**Comment by nickrandolph** [comment]
Any update on this issue?

**Comment by bgavrilMS** [comment]
We found some problems (missing API) on mobile, which would prevent system browser and broker scenarios. We are discussing this with Maui team. 

On a positive note, I tested WinUI and it works well with MSAL 4.44 (already released). No changes needed. 

@SameerK-MSFT - please keep this thread updated with progress / blockers etc. Some of the folks here might be able to help.

**Comment by jeromelaban** [comment]
@bgavrilMS could you tell which API is missing? Is it an API specific to MAUI or to `net6.0-ios` or `net6.0-android` ?

**Comment by bgavrilMS** [comment]
Following API is missing from AppDelegate on iOS.

```csharp
public override bool OpenUrl(UIApplication app, NSUrl url, string sourceApplication, NSObject annotation)
```

This hook is implemeted by the app developer like [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/tests/devapps/Xamarin%20Manual/Xamarin%20Manual/Xamarin%20Manual.iOS/AppDelegate.cs#L34). MSAL uses to this to know if the broker was used or not - see [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/09cd50a3027b811926f8ad53c4b18e287c8b9294/src/client/Microsoft.Identity.Client/Platforms/iOS/AuthenticationContinuationHelper.cs#L42). We haven't looked into this too much, as this `sourceApplication` param isn't anyway supported on iOS 13+ (it's empty), so maybe it's not an issue to always ignore it? 

The Maui team recommended we use https://gist.github.com/Redth/5be697ac7e75e8d47c26deea83fbc3d0 

@SameerK-MSFT is working on this, so pls take with a grain of salt...

**Comment by jeromelaban** [comment]
Thanks. To clarify, are your new changes still going to be including a non-maui `net6.0-ios` package?

**Comment by bgavrilMS** [comment]
What do you mean by non-maui "net6-ios" package? I thought MAUI means `net6-ios`, `net6-android`, `net6-maccatalyst`

Note that MSAL fully supports the vanilla net6 target, e.g. confidential client scenarios and public client desktop scenarios (including WinUI).

**Comment by bgavrilMS** [comment]
> > Following API is missing from AppDelegate on iOS.
> > ```cs
> > public override bool OpenUrl(UIApplication app, NSUrl url, string sourceApplication, NSObject annotation)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This hook is implemeted by the app developer like [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/tests/devapps/Xamarin%20Manual/Xamarin%20Manual/Xamarin%20Manual.iOS/AppDelegate.cs#L34). MSAL uses to this to know if the broker was used or not - see [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/09cd50a3027b811926f8ad53c4b18e287c8b9294/src/client/Microsoft.Identity.Client/Platforms/iOS/AuthenticationContinuationHelper.cs#L42). We haven't looked into this too much, as this `sourceApplication` param isn't anyway supported on iOS 13+ (it's empty), so maybe it's not an issue to always ignore it?
> > The Maui team recommended we use https://gist.github.com/Redth/5be697ac7e75e8d47c26deea83fbc3d0
> > @SameerK-MSFT is working on this, so pls take my words with a grain of salt...
> 
> it exists: https://github.com/dotnet/maui/blob/0872cbc326ea8230c24c075d2f98b3195df69765/src/Core/src/Platform/iOS/MauiUIApplicationDelegate.cs#L80

Notice how it does not have the `string sourceApplication` param (which was deprecated by Apple in ios 13)

**Comment by bgavrilMS** [comment]
> MAUI is a layer on top of the Android/iOS/Catalyst bindings (`net6.0-android`, `net6.0-ios` and others). Making a "native" (non-MAUI) app is possible (e.g. [Uno Platform](https://github.com/unoplatform/uno) uses that specific part) and if you add an explicit dependency to MAUI nuget packages in MSAL, this will become breaking for those frameworks and apps.

The only dependency I see are: 

```
<PackageReference Include="System.Security.SecureString" Version="4.3.0" />
<PackageReference Include="Xamarin.AndroidX.Browser" Version="1.4.0" PrivateAssets="All" />
```

(not sure why we depdend on SecureString, I think it needs to go)

**Comment by jeromelaban** [comment]
> The only dependencies I see are:
> 
> ```
> <PackageReference Include="System.Security.SecureString" Version="4.3.0" />
> <PackageReference Include="Xamarin.AndroidX.Browser" Version="1.4.0" PrivateAssets="All" />
> ```
> 
> (not sure why we depdend on SecureString, I think it needs to go)

This should be good then, thank you! I was mentioning this because of the changes that would be required for the appdelegate, but those are likely to be documentation changes.

**Comment by joker109** [comment]
@bgavrilMS 
Is it working in iOS by using the MauiBuilder extension method as recommended by the MAUI team.
<img width="833" alt="Screenshot 2022-06-22 at 4 49 11 PM" src="https://user-images.githubusercontent.com/34001302/175016709-38bbd9cc-2b15-4cff-9be7-6528868c4f0b.png">

**Comment by kzryzstof** [comment]
This comment is a follow-up to changes I observed after upgrading the package to 4.45 for my `net6-ios` application (**non-MAUI**). 

I used to have a hard crash when I used 4.43 as described in [the issue 3272](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3272).

Now I get a surprising exception related to the redirect URI parameter:

```
MSAL.NetCore.4.45.0.0.MsalClientException: 
	ErrorCode: loopback_redirect_uri
Microsoft.Identity.Client.MsalClientException: Only loopback redirect uri is supported, but **msal<app_id>//auth/** was found. Configure http://localhost or http://localhost:port both during app registration and when you create the PublicClientApplication object. See https://aka.ms/msal-net-os-browser for details
   at Microsoft.Identity.Client.Platforms.Shared.Desktop.OsBrowser.DefaultOsBrowserWebUi.UpdateRedirectUri(Uri )
   at Microsoft.Identity.Client.Internal.AuthCodeRequestComponent.FetchAuthCodeAndPkceInternalAsync(IWebUI , CancellationToken )
   at Microsoft.Identity.Client.Internal.AuthCodeRequestComponent.FetchAuthCodeAndPkceVerifierAsync(CancellationToken )
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.GetTokenResponseAsync(CancellationToken )
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken )
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken )
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters , AcquireTokenInteractiveParameters , CancellationToken )
```
Since I am using Azure B2C, I can't use localhost as it is not supported.

But, what makes this error even more suspicious is that the static class `AuthenticationContinuationHelper` can't be found even though the `Microsoft.Identity.Client` NuGet package is referenced:

<img width="1324" alt="Screen Shot 2022-06-26 at 9 17 40 AM" src="https://user-images.githubusercontent.com/38137839/175816014-83ce1eb9-cc50-466b-bb28-897f2cac3062.png">

With the current state of things related to the issue and the changes made, is that expected?

And to make it clear like others did, I am **not** using MAUI. It is a `xamarin.ios` application that been migrated to `.net6-ios` :)

**Comment by bgavrilMS** [comment]
@kzryzstof - since we haven't managed to add net6-ios support to MSAL, nuget will choose MSAL's netcore2.1 implementation in your app. In this implementation we use a desktop system browser (based on http://localhost redirect uri). So that explains the error and why `AuthenticationContinuationHelper` does not work.

**Comment by joker109** [comment]
We need to use Azure AD authentication in our .net MAUI app targeting iOS. Is there any alternative way to achieve that without using MSAL?

We are on a tight schedule. Any help is appreciated.

**Comment by RichardMarsh** [comment]
> We need to use Azure AD authentication in our .net MAUI app targeting iOS. Is there any alternative way to achieve that without using MSAL?
> 
> We are on a tight schedule. Any help is appreciated.

Same here. Any feedback or timeline on this would be greatly appreciated.

**Comment by xperiandri** [comment]
> Is there any alternative way to achieve that without using MSAL?

https://github.com/IdentityModel/IdentityModel.OidcClient

**Comment by VladislavAntonyuk** [comment]
MAUI has WebAuthenticator which you can use to achieve the same.

**Comment by mrDenning** [comment]
> MAUI has WebAuthenticator which you can use to achieve the same.

Do you have an example / link?

**Comment by anpin** [comment]
> MAUI has WebAuthenticator which you can use to achieve the same.

But it doesn't work on Windows yet https://github.com/dotnet/maui/issues/2702 

Since MAUI is GA we need solution for all target platforms

**Comment by SameerK-MSFT** [comment]
sameerk/Maui_CI branch now supports iOS and Android. **Note: this branch is still in development. ** It has two devapp samples (i.e. the samples directly reference the project). There is no nuget package. It has not been tested with other platforms. The instructions to run the samples are provided **[here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/sameerk/Maui_CI/MauiStatus.md)**. Please check and let us know how it goes.

**Comment by nickrandolph** [comment]
@SameerK-MSFT can you please include sample net6-iOS and net6-android applications that are not Maui applications - this library should be agnostic from the UI framework, so it'd be clearer to have samples that demonstrate that it can be used in non-Maui scenarios.

**Comment by bgavrilMS** [comment]
@nickrandolph - we will leave this as an exercise for you. MSAL does not reference anything that is UI specific. I am not sure how popular non-Maui applications, for now we are targetting Maui. I think one of our internal partners already tried stand-alone ios and it worked fine.

**Comment by nickrandolph** [comment]
@bgavrilMS this makes no sense - Maui is a UI library that sits on top of .NET for iOS/Android/WinUI. It would be better for dev/testing of the library to be done without taking the dependency on Maui. There are other stable frameworks such as Avalonia and Uno that are will rely on this capabilities (and are currently blocked from shipping on net6 because of the lack of support!)

**Comment by bgavrilMS** [comment]
Thanks @nickrandolph. We are not mobile experts, but my hypothesis is that MAUI is the first priority. MSAL will not take any dependencies on MAUI or on UI, except for the browser component on Android (see the PR associated, look at the csproj file). We do not have the capacity of testing Avalonia and Uno but will prioritize any bugs you may find.

**Comment by bgavrilMS** [comment]
We'd like to have a preview package out by end of this week, hopefully in a day or two.

**Comment by mrDenning** [comment]
Does the preview release released today include this fix / feature?

**Comment by joker109** [comment]
<img width="902" alt="Screenshot 2022-07-08 at 9 52 48 AM" src="https://user-images.githubusercontent.com/34001302/177916516-561bd039-8623-47bd-8057-b9c362116140.png">
@mrDenning , We are not able to find any preview release yet.

> Does the preview release released today include this fix / feature?

**Comment by bgavrilMS** [comment]
We are still working on this and have not released a preview. Getting our CI to build with these new targets is proving challenging. The PR has been merged; we need to do some testing before pushing it to nuget.

**Comment by SameerK-MSFT** [comment]
OK. The preview release "4.46.0-preview2" is out on the NuGet. Also, you can find more information about how to use it, known limitations etc [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/MauiStatus.md)

**Comment by mrDenning** [comment]
Great news! Will try it out and provide feedback

**Comment by SameerK-MSFT** [comment]
[nickrandolph](https://github.com/nickrandolph) The Xamarin.Forms samples are in these branches.
https://github.com/Azure-Samples/active-directory-xamarin-native-v2/tree/sameerk/Validate_Maui_Preview
https://github.com/Azure-Samples/active-directory-b2c-xamarin-native/tree/sameerk/ValidateOnMaui

**Comment by kzryzstof** [comment]
@SameerK-MSFT @bgavrilMS I am sooo happy to confirm that my iOS application (`.net6-ios`) is now working again (at least of the iOS simulator). Thank you very much :)

**Comment by anpin** [comment]
Using `4.46.0-preview2` in `MAUI` `net6.0-windows` target do I need to enable anything for cache to work? Every time I start the app in debug mode the cache is empty and I have to login again. I'm yet to try the packaged release version, but my `launchProfile` is using the `MsixPackage` command, so I suppose it should be the same. The app in question is ported over from `Xamarin.Forms` where `UWP` target was working perfectly fine in `Debug`. Here is my initialization snippet
```
var handler = PublicClientApplicationBuilder.Create(AzureConstants.ClientID)
            .WithB2CAuthority(AzureConstants.AuthoritySuSi)
#if __IOS__ || __MACCATALYST__
                            .WithIosKeychainSecurityGroup("com.microsoft.adalcache")
#endif
            .WithRedirectUri(AzureConstants.RedirectUri)
            .Build();

```
With logging enabled I see this on startup
```
[15:35:18 FTL] False MSAL 4.46.0.0 MSAL.NetCore .NET 6.0.5 Microsoft Windows 10.0.19044 [2022-07-09 11:35:18Z] [GetAllAccessTokens] Total number of cache partitions found while getting refresh tokens: 0
[15:35:18 INF] False MSAL 4.46.0.0 MSAL.NetCore .NET 6.0.5 Microsoft Windows 10.0.19044 [2022-07-09 11:35:18Z - GUID] Skipping Instance discovery for B2C authority
[15:35:18 INF] False MSAL 4.46.0.0 MSAL.NetCore .NET 6.0.5 Microsoft Windows 10.0.19044 [2022-07-09 11:35:18Z - GUID] [Instance Discovery] Skipping Instance discovery for non-AAD authority. 
[15:35:18 INF] False MSAL 4.46.0.0 MSAL.NetCore .NET 6.0.5 Microsoft Windows 10.0.19044 [2022-07-09 11:35:18Z] Found 0 cache accounts and 0 broker accounts
[15:35:18 INF] False MSAL 4.46.0.0 MSAL.NetCore .NET 6.0.5 Microsoft Windows 10.0.19044 [2022-07-09 11:35:18Z] Returning 0 accounts
```

**Comment by chrfin** [comment]
@anpin Did you add the \`Microsoft.Identity.Client.Extensions.Msal \` package and enabled the cache via a \`MsalCacheHelper \`?

**Comment by nickrandolph** [comment]
@chrfin is this going to continue to be the way to persist tokens? Will persistence be added to the core MSAL package for all supported platforms?

**Comment by chrfin** [comment]
@nickrandolph I do not know, I just know I needed to do this for MAUI on Windows, when I implemented a prototype some month ago...  

Maybe @SameerK-MSFT can answer this?

**Comment by anpin** [comment]
@chrfin thank you it works with the mentioned package now, however I'm pretty sure it was working without it in the legacy XF /UWP project

**Comment by billyjacobs2014** [comment]
I am using the latest preview version (4.46.0-preview2). 

I have an Entitlements.plist file in my platforms/ios folder which contains the following:

`<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>aps-environment</key>
		<string>development</string>
		<key>keychain-access-groups</key>
		<array>
			<string>$(AppIdentifierPrefix)com.microsoft.adalcache</string>
		</array>
	</dict>
</plist>`

The code to acquire a token is contained here: 
`    public class AuthService
    {
        private readonly IPublicClientApplication authenticationClient;
        public AuthService()
        {
            authenticationClient = PublicClientApplicationBuilder
                .Create(AzureADConstants.ClientId)
                //.WithB2CAuthority(Constants.AuthoritySignIn) // uncomment to support B2C
                .WithAuthority("https://login.microsoftonline.com/35347e8b-7fa3-4f36-9853-da58952dbc50/oauth2/v2.0/authorize")
                .WithIosKeychainSecurityGroup("com.microsoft.adalcache")
                .WithRedirectUri($"msal{AzureADConstants.ClientId}://auth")
                .Build();
            // msalbdd7cd39-b363-4cdd-b1b2-6dab6695be7b://auth
        }

        public async Task<AuthenticationResult> LoginAsync(CancellationToken cancellationToken)
        {
            try
            {
                var result = await authenticationClient
                    .AcquireTokenInteractive(AzureADConstants.Scopes)
                    .WithPrompt(Prompt.ForceLogin)
#if ANDROID
                    .WithParentActivityOrWindow(Platform.CurrentActivity)
#endif
                    .ExecuteAsync(cancellationToken);
                return result;
            }
            catch (MsalClientException ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.Message);
                return null;
            }
        }
    }`

Good news is that now everything is building, and deploys to my remote ios simulator, and the login screen appears correctly. 

Bad news is that after I login, I get an MsalClientException. Below is the call stack. I searched this error and navigated all of the github move links which finally brought me here. 

`ErrorCode: missing_entitlements
Microsoft.Identity.Client.MsalClientException: The application does not have keychain access groups enabled in the Entitlements.plist. As a result, there was a failure to save to the iOS keychain. The keychain access group 'com.com.microsoft.adalcache' is not enabled in the Entitlements.plist. Also, use the WithIosKeychainSecurityGroup api to set the keychain access group. See https://aka.ms/msal-net-enable-keychain-groups for more details on enabling keychain access groups and entitlements.
   at Microsoft.Identity.Client.Platforms.iOS.iOSTokenCacheAccessor.Save(IiOSKey key, String payload)
   at Microsoft.Identity.Client.Platforms.iOS.iOSTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item)
   at Microsoft.Identity.Client.TokenCache.Microsoft.Identity.Client.ITokenCacheInternal.SaveTokenResponseAsync(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.TokenCache.Microsoft.Identity.Client.ITokenCacheInternal.SaveTokenResponseAsync(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.Cache.CacheSessionManager.SaveTokenResponseAsync(MsalTokenResponse tokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.CacheTokenResponseAndCreateAuthenticationResultAsync(MsalTokenResponse msalTokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenInteractiveParameters interactiveParameters, CancellationToken cancellationToken)`

**Comment by NPadrutt** [comment]
@billyjacobs2014  I had the same. For me it worked, once I copied over the following part from my Xamarin.Forms iOS project to register the Entitlements correctly


```
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|iPhoneSimulator' ">
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|iPhoneSimulator' ">
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|iPhone' ">
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|iPhone' ">
    <CodesignEntitlements>Entitlements.plist</CodesignEntitlements>
  </PropertyGroup>
```

**Comment by joker109** [comment]
@billyjacobs2014, 
Is it working for you? I added the tags for Entitlements.plist in project file as suggested by @NPadrutt. But still its not working for me.

**Comment by VladislavAntonyuk** [comment]
It should look like this: https://github.com/VladislavAntonyuk/MauiSamples/blob/main/Directory.Build.props#L25

**Comment by bgavrilMS** [comment]
> 

For the Windows target, you have to add caching manually. This is due to an unfortunate decision some time ago, to not provide persisted caching out of the box. Anyway - doc is here: https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-token-cache-serialization?tabs=aspnetcore

I also recommend folks to upgrade to use brokers on both Windows and mobile. They are better at SSO and better at security.

**Comment by bgavrilMS** [comment]
To troubleshoot entitlements and KeyChain access on ios, please see this short doc I created a while back: 

https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-ios-considerations#troubleshooting-keychain-access

**Comment by lrgar** [comment]
> To troubleshoot entitlements and KeyChain access on ios, please see this short doc I created a while back. The gist of it is - your app must be able to access the KeyChain on its own, because MSAL needs KeyChain access to store tokens and other data there.
>
> https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-ios-considerations#troubleshooting-keychain-access

I got into this situation actually, and in case it helps anyone: the [docs](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-ios-considerations#troubleshooting-keychain-access) have an example attempting to write into KeyChain to confirm access. I did have access, that wasn't the problem.

The [MSAL library](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/90958ca7c420933a36bae6877e740bdeb9396534/src/client/Microsoft.Identity.Client/Platforms/iOS/iOSTokenCacheAccessor.cs#L44) does something similar, but, when it tried to query for the `DotNetTeamIDHint` key, it'd get an `ItemNotFound` error. It then tried to insert it and was getting a `DuplicateItem` error on the `SecKeyChain.Add()` call. It was only with this key, as I tried setting and querying other keys without issue.

This was with the iOS simulator, after I upgraded from a local copy of the repository with the aforementioned PR branch to the MSAL preview version in NuGet. But I presume it was a bug with the simulator itself which I solved by resetting the simulator data. It worked fine after that.

**Comment by anpin** [comment]
@bgavrilMS can you prove how brokers are better in security? For now it seems like a pure marketing statement to push `Microsoft Authenticator` app to the end-users through yet another channel.

**Comment by bgavrilMS** [comment]
@anpin - there is an immediate benefit and a longer-term plan in terms of security.

Immediate benefit for using a broker is that the application is not given a refresh token (which is a very powerful artefact - 3-month expiration, can be used to get access tokens for any resource the user has consented to). The broker handles all refresh tokens internally and the tokens are hardware bound where possible (see https://docs.microsoft.com/en-us/azure/active-directory/devices/concept-primary-refresh-token).

Longer-term plan is that we want to move away from Bearer access tokens to bounded access tokens. The broker takes care of the RT (i.e. RT will be bound). The broker will then help the application bind the access token as well. See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Proof-Of-Possession-(PoP)-tokens

To sum up: with broker, RT is not given to apps and is cryptographically bound to the device (where possible). we are working on doing the same for AT.

**Comment by anpin** [comment]
@bgavrilMS Why can't tokens be hardware bound with this library? According to [this](https://docs.microsoft.com/en-us/azure/active-directory-b2c/configure-tokens?pivots=b2c-user-flow) in B2C one can customize refresh token lifetime as they like and I'm sure other OAuth providers would have this option as well. However I still don't get how trusting your "very powerful artefact" to a 3rd party closed source app is considered more secure, maybe you could correct me if I'm wrong here. It also adds additional level of complexity which I personally tend to avoid.  One example of how it could ruin app experience was immediately found in AppStore reviews for MS Authenticator 
![image](https://user-images.githubusercontent.com/6060545/179445896-b4739edf-6923-46a9-92f5-d424f4af5c0e.png)

**Comment by bgavrilMS** [comment]
Authenticator / brokers do not exist for B2C. We are only talking about AAD auth, i.e. Work and School accounts and Personal Microsoft accounts (e.g. outlook.com.

They help with token binding and also with SSO (i.e. users don't have to re-enter passwords). If tenant admin enables certain Conditional Access policies such as "device must be managed", intune management etc. then Authenticator helps prove these claims. So it's better if app is configured to use Authenticator, and if tenant admin enables these policies, app doesn't have to change.

Architecturally, it makes sense to push some functionality to Authenticator, because it can be updated easily for the benefit of all apps. We have found that most app developers do not update their apps to consume latest identity SDK, but Authenticator gets updated via app store. 

This is good feedback, I'll pass it along. If you wish to discuss more on this topic, is it ok to 
move to the Discussions page, so as to leave the thread for Maui issues?

**Comment by bgavrilMS** [comment]
@SameerK-MSFT  - can you please create individual work items for the remaining work? Ideally break them down into ~2 day tasks.

**Comment by SophisticatedConsulting** [comment]
When MAUI MSAL support be released... 2 new versions 4.46.0 and 4.46.1 seem to be without MAUI support?

**Comment by bgavrilMS** [comment]
@SophisticatedConsulting - the preview versions have MAUI support. We're having problems with our CI pipelines, which are mainly geared towards services, to get all the SDKs in one place etc. Not anticipating any functional changes but waiting for CI agents to be upgraded.

**Comment by kzryzstof** [comment]
Just to be clear, anyone using MAUI or net-ios should stay on version 4.46-preview2 until further notice. Is that correct?

Téléchargez Outlook pour iOS<https://aka.ms/o0ukef>
________________________________
De : Bogdan Gavril ***@***.***>
Envoyé : Monday, August 22, 2022 6:56:58 AM
À : AzureAD/microsoft-authentication-library-for-dotnet ***@***.***>
Cc : Christophe Commeyne ***@***.***>; Mention ***@***.***>
Objet : Re: [AzureAD/microsoft-authentication-library-for-dotnet] [Feature Request] Make MSAL.NET compatible with MAUI (Issue #3127)


@SophisticatedConsulting<https://github.com/SophisticatedConsulting> - the preview versions have MAUI support. We're having problems with our CI pipelines, which are mainly geared towards services, to get all the SDKs in one place etc. Not anticipating any functional changes but waiting for CI agents to be upgraded.

**Comment by bgavrilMS** [comment]
Yes, @kzryzstof, this is correct.

If we make any changes that affect mobile flows, we will release a new preview version.

**Comment by Orgbrat** [comment]
So just where do I find the 4.46-preview2 so that I may use it within my MAUI application.  All I can get to show in Visual Studio is version 4.46.1 even with "include pre-release" checked.

**Comment by tipa** [comment]
@Orgbrat likely just a UI bug in VS. You can hardcode it in your .csproj.
The package is available: https://www.nuget.org/packages/Microsoft.Identity.Client/4.46.0-preview2

**Comment by DRAirey1** [comment]
> NOTE: One thing to notice in the error is that it says **com.com.microsoft.adalcache** in the error where com is repeated.

I'm running into this exact issue. None of the answers here has helped. When I add this to the project file:

```
  <PropertyGroup>
          ....
    <CodesignEntitlement>MyEntitlements.plist</CodesignEntitlement>
  </PropertyGroup>

```
And add a file called 'MyEntitlements.plist', then I get no error messages, but the application crashes with this error:


So then I take the 'CodeSignEntitlements' out of the project file, I get the same error. It doesn't appear like the 'CodesignEntitlements' element is honored in the build. So then I just rename the file to 'Entitlements.plist' and I get the missing_entitlements error sited [https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3127#issuecomment-1180457402](url).

So, to sum up, I'm running into the exact same issue sited in that link, and nothing I do with the <CodesignEntitlement> element or the 'Entitlements.plist' seems to change anything.

**Comment by igalfsg** [comment]
Hi I just saw that Mac Catalyst was cut, is there a plan to support it in the future?

**Comment by DRAirey1** [comment]
To anyone having this same problem, here's a workaround until they fix these bugs:

1. Visual Studio doesn't care about the **CodesignEntitlement** element in the project file, it only cares about a file named 'Entitlements.plist' existing in the iOS directory of a Maui project.
2. The MSAL package appears to be unable to find valid profiles, so it can't expand the $(AppIdentifierPrefix). However, you can hard-code your Team ID prefix into the 'Entitlements.plist' file like so:

```
<plist version="1.0">
  <dict>
    <key>keychain-access-groups</key>
    <array>
      <string>8BR42YBXJD.com.microsoft.adalcache</string>
    </array>
  </dict>
</plist>

``` 
Authentication appears to work after doing this, but there are several non-fatal exceptions that are introduced after doing this, like the IStringLocalizer is unable to find the resource files, but then continues working after the exception is thrown.

**Comment by DRAirey1** [comment]
> @joker109 It should look like this: https://github.com/VladislavAntonyuk/MauiSamples/blob/main/Directory.Build.props#L25

Sorry, but all evidence is that Maui projects don't pay any attention to the settings in the project file when looking for the Enlistments.plist. It doesn't even look like you can rename the file. It's hard-coded.

**Comment by bgavrilMS** [comment]
> Hi I just saw that Mac Catalyst was cut, is there a plan to support it in the future?

CC @oldalton for requests to support MSAL on Mac Catalyst.

**Comment by kzryzstof** [comment]
Aucun problème.

________________________________________
De : Bogdan Gavril ***@***.***>
Envoyé : Monday, August 29, 2022 7:22:52 AM
À : AzureAD/microsoft-authentication-library-for-dotnet
Cc : Christophe Commeyne; Mention
Objet : Re: [AzureAD/microsoft-authentication-library-for-dotnet] [Feature Request] Make MSAL.NET compatible with MAUI  (Issue #3127)

Hi I just saw that Mac Catalyst was cut, is there a plan to support it in the future?

CC @oldalton<https://github.com/oldalton> for requests to support MSAL on Mac Catalyst.

**Comment by Orgbrat** [comment]
Microsoft has gone full in on MAUI and in my opinion that means that all supporting Microsoft libraries should also go full in on their Maui support. You should not be cutting platforms but making the library support all the Microsoft built-in platforms at the very least. So please, do not drop Mac Catalyst but figure out how to get it working.

Please, please and thanks...

**Comment by SophisticatedConsulting** [comment]
Please support Mac Catalyst.

**Comment by bgavrilMS** [comment]
Root cause for not supporting MacCatalyst - https://github.com/AzureAD/microsoft-authentication-library-for-objc/issues/649

Basically we're not sure how to store the token cache so that it doesn't break future scenarios. We could let app developers store the token cache, similar to how we do on Windows (WinUI). Thoughts?

**Comment by nickrandolph** [comment]
@bgavrilMS you mean letting developers provide their own token cache?

**Comment by Ruvi1996** [comment]
So now I see that the 4.47 has been released with the Maui support. 
But the issue is still there. 

Not only in my project. [The sample](https://github.com/Azure-Samples/active-directory-xamarin-native-v2) given in Azure-Samples repository is also getting this error. That sample also didn't have entries @DRAirey1 have mentioned above so I added them. But still getting that same error on that project too.

Here is the error (this is from the MauiB2C project from the sample mentioned above even after adding the CodesEntitlements entries):

```
ObjCRuntime.ObjCException
  Message=Objective-C exception thrown.  Name: Microsoft.Identity.Client.MsalClientException Reason: The application does not have keychain access groups enabled in the Entitlements.plist. As a result, there was a failure to save to the iOS keychain. The keychain access group 'com.com.microsoft.adalcache' is not enabled in the Entitlements.plist. Also, use the WithIosKeychainSecurityGroup api to set the keychain access group. See https://aka.ms/msal-net-enable-keychain-groups for more details on enabling keychain access groups and entitlements. (Microsoft.Identity.Client.MsalClientException)
   at Microsoft.Identity.Client.Platforms.iOS.iOSTokenCacheAccessor.Save(IiOSKey key, String payload)
   at Microsoft.Identity.Client.Platforms.iOS.iOSTokenCacheAccessor.SaveAccessToken(MsalAccessTokenCacheItem item)
   at Microsoft.Identity.Client.TokenCache.Microsoft.Identity.Client.ITokenCacheInternal.SaveTokenResponseAsync(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.TokenCache.Microsoft.Identity.Client.ITokenCacheInternal.SaveTokenResponseAsync(AuthenticationRequestParameters requestParams, MsalTokenResponse response)
   at Microsoft.Identity.Client.Cache.CacheSessionManager.SaveTokenResponseAsync(MsalTokenResponse tokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.CacheTokenResponseAndCreateAuthenticationResultAsync(MsalTokenResponse msalTokenResponse)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenInteractiveParameters interactiveParameters, CancellationToken cancellationToken)
   at MauiB2C.MSALClient.PCAWrapperB2C.AcquireTokenInteractiveAsync(String[] scopes) in C:\Users\RuvindraSenevirathne\source\repos\active-directory-xamarin-native-v2\MAUI\MauiAppB2C\MSALClient\PCAWrapperB2C.cs:line 65
   at MauiB2C.MainPage.OnSignInClicked(Object sender, EventArgs e) in C:\Users\RuvindraSenevirathne\source\repos\active-directory-xamarin-native-v2\MAUI\MauiAppB2C\MainPage.xaml.cs:line 32
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_0(Object state)
   at Foundation.NSAsyncSynchronizationContextDispatcher.Apply()

Native stack trace:
	0   CoreFoundation                      0x0000000110314604 __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x000000011d254a45 objc_exception_throw + 48
	2   libxamarin-dotnet-debug.dylib       0x000000010bd8addf xamarin_process_managed_exception + 943
	3   MauiB2C                             0x0000000102d65c19 _ZL31native_to_managed_trampoline_10P11objc_objectP13objc_selectorPP11_MonoMethodj + 361
	4   MauiB2C                             0x0000000102d663f9 -[__MonoMac_NSAsyncSynchronizationContextDispatcher xamarinApplySelector] + 41
	5   Foundation                          0x00000001133055d7 __NSThreadPerformPerform + 179
	6   CoreFoundation                      0x00000001102814a7 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
	7   CoreFoundation                      0x000000011028139f __CFRunLoopDoSource0 + 180
	8   CoreFoundation                      0x000000011028086c __CFRunLoopDoSources0 + 242
	9   CoreFoundation                      0x000000011027af68 __CFRunLoopRun + 871
	10  CoreFoundation                      0x000000011027a704 CFRunLoopRunSpecific + 562
	11  GraphicsServices                    0x000000011e563c8e GSEventRunModal + 139
	12  UIKitCore                           0x0000000128b5665a -[UIApplication _run] + 928
	13  UIKitCore                           0x0000000128b5b2b5 UIApplicationMain + 101
	14  libmonosgen-2.0.dylib               0x000000010c578aa2 do_icall + 194
	15  libmonosgen-2.0.dylib               0x000000010c577abd do_icall_wrapper + 253
	16  libmonosgen-2.0.dylib               0x000000010c569cda interp_exec_method + 2970
	17  libmonosgen-2.0.dylib               0x000000010c567f5f interp_runtime_invoke + 239
	18  libmonosgen-2.0.dylib               0x000000010c45fbab mono_jit_runtime_invoke + 1227
	19  libmonosgen-2.0.dylib               0x000000010c382198 mono_runtime_invoke_checked + 136
	20  libmonosgen-2.0.dylib               0x000000010c388f3b mono_runtime_exec_main_checked + 107
	21  libmonosgen-2.0.dylib               0x000000010c4c2042 mono_jit_exec + 354
	22  libxamarin-dotnet-debug.dylib       0x000000010bd9ee9d xamarin_main + 1949
	23  MauiB2C                             0x0000000102df7844 main + 52
	24  dyld                                0x000000010b972f21 start_sim + 10
	25  ???                                 0x000000020340452e 0x0 + 8644478254
	26  ???                                 0x0000000000000003 0x0 + 3

  Source=Microsoft.iOS
  StackTrace:
   at ObjCRuntime.Runtime.ThrowNSException(IntPtr ns_exception)
   at ObjCRuntime.Runtime.throw_ns_exception(IntPtr exc)
   at UIKit.UIApplication.Main(String[] args, Type principalClass, Type delegateClass)
   at MauiB2C.Program.Main(String[] args) in C:\Users\RuvindraSenevirathne\source\repos\active-directory-xamarin-native-v2\MAUI\MauiAppB2C\Platforms\iOS\Program.cs:line 15
```

**Comment by bgavrilMS** [comment]
Is this the iOS app @Ruvi1996 ? Did you try the troubleshooting steps at https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-xamarin-ios-considerations#troubleshooting-keychain-access ? Basically - can your app write to the KeyChain?

**Comment by Ruvi1996** [comment]
@bgavrilMS  This is a MAUI project. And yes I'm debugging the IOS app.
I tried the troubleshooting steps and I got the result as success so I believe my app can write to the KeyChain.

![image](https://user-images.githubusercontent.com/59006361/192195344-b54f3830-dbcc-4246-8ccf-fbc6a0bac202.png)

**Comment by bgavrilMS** [comment]
@Ruvi1996 - do you mind creating a new bug for the KeyChain access issue please? This thread is too long and the actual feature has been completed.

**Comment by AathifMahir** [comment]
@bgavrilMS Does Token Caching for Local Storage Works with the current release, as Testing Maui App on Windows, I don't see Msal.net Stores Token in the Storage, every time we exit the app, we lose the token

**Comment by bgavrilMS** [comment]
Hi @AathifMahir - you are right, we don't cache the tokens on WinUI app on Windows, only on mobile. We have a work item to update the sample on how to do this, but really the instructions here should be clear: https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-token-cache-serialization?tabs=desktop

i.e. you'll store the token cache in a file encrypted on disk and we have a library for that.

**Comment by AathifMahir** [comment]
> Hi @AathifMahir - you are right, we don't cache the tokens on WinUI app on Windows, only on mobile. We have a work item to update the sample on how to do this, but really the instructions here should be clear: https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-token-cache-serialization?tabs=desktop
> 
> i.e. you'll store the token cache in a file encrypted on disk and we have a library for that.

What about MacCatalyst Maui? Does it have a built in token cache?

**Comment by bgavrilMS** [comment]
We don't support MacCatalyst

**Comment by bgavrilMS** [comment]
@SameerK-MSFT will know more about experience on MacCatalyst

**Comment by igalfsg** [comment]
@bgavrilMS and @SameerK-MSFT  is there any update on the MacCatalyst support? Here is how auth0 is doing it, hope it helps :) https://github.com/auth0-blog/dotnet-blazor-maui-auth0-app

**Comment by bgavrilMS** [comment]
Re-opening this so as to help prioritize. CC @localden and @davidortinau

I want to share how to get MSAL to work on .net Mac and on .net MacCatalyst through [extensiblity points](https://learn.microsoft.com/en-us/entra/msal/dotnet/advanced/extensibility-points). MSAL handles 2 aspects which are OS specific: token caching and browser pop-up which will need implementing. 

Note that an app targeting MacCatalyst or MacOS via netX-mac / netX-maccatalyst will cause NuGet to resolve to MSAL's netstadard target framework. Here, MSAL's behavior is:

- in-memory token caching only. Tokens are not persisted. Restarting the app means user needs to re-auth.
- a system browser is kicked off. Won't work on iOS. 

### Token Caching

`KeyChain` is the preferred storage for secrets on iOS and macOS. 
Extensibility Point: https://learn.microsoft.com/en-us/entra/msal/dotnet/how-to/token-cache-serialization?tabs=custom
Suggested implementation: Use the KeyChain APIs to persist the token cache (a byte array) in the KeyChain. [Example usage](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/Platforms/iOS/iOSLegacyCachePersistance.cs) of KeyChain APIs.

### Web Browser 

Extensibility Point: https://learn.microsoft.com/en-us/entra/msal/dotnet/advanced/extensibility-points#desktop--mobile-apps---icustomwebui
Suggested Implementiation: You can draw inspiration from the net6-iOS [system browser](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/Platforms/iOS/SystemWebview/SystemWebUI.cs) or [embedded browser](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/Platforms/iOS/EmbeddedWebview/EmbeddedWebUI.cs)  implementations

**Comment by igalfsg** [comment]
was MacOS completed? or closed as won't fix again?

**Comment by bgavrilMS** [comment]
Ah, no, MacOs support is not there yet, but there is work being done on this.

**Comment by moilotus** [comment]
Are we therefore expecting a solution to macOS Catalyst in the near future for MSAL?

I'm trying to do auth on PC/macOS/iOS/Android for MAUI.
WebAuthenticator can do iOS/macOS/Android, but fails on PC. They helpfully tell us on the same page, and I quote
"Tip

[Microsoft Authentication Library (MSAL)](https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-overview) provides an excellent turn-key solution to adding authentication to your app."

But it would appear that whilst MSAL can do PC/iOS/Android, it fails on macOS?

So currently neither MS auth system out of the box works for all four major MAUI  platforms, or am I missing something?

Many thanks.

**Comment by moilotus** [comment]
Many thanks for the reply. In the end I went with a workaround for WebAuthenticator not working on PC instead of using MSAL. Nice to know there are some workarounds.


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by jeromelaban** [comment]
@bgavrilMS It's a odd choice to prioritize appkit over catalyst, mainly because MAUI does not support it, and that the build tooling appkit is not very stable at this time (as it's not supported by MAUI).

Thanks for the workarounds.

**Comment by bgavrilMS** [comment]
> @bgavrilMS It's a odd choice to prioritize appkit over catalyst, mainly because MAUI does not support it, and that the build tooling appkit is not very stable at this time (as it's not supported by MAUI).
> 
> Thanks for the workarounds.

There is a technical limitation with token caching on MacCatalyst, related to KeyChain access, which today would make it so that apps don't attest to device compliance (this is an AAD only concept, not present in B2C). So not really a choice.

**Comment by mcmcelro** [comment]
Hi - I'm not sure if the team knows this or not, but the file that's supposed to list the current status appears to be missing: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/MauiStatus.md

**Comment by localden** [comment]
Closing this issue as MAUI is supported in MSAL.NET. Example covered [in documentation](https://learn.microsoft.com/azure/developer/mobile-apps/azure-mobile-apps/quickstarts/maui/authentication).
